### PR TITLE
Commissioner discovery by android tv-casting-app

### PR DESCRIPTION
**Pull request by sharadb-amazon**
#### Problem
The Android version of the tv-casting-app needs to be able to discover Matter commissioners using DNS-SD.

#### Change overview
Using the Android NSDManager to discover Matter commissioners and display them on the app UI. 

#### Testing
* Unit tests added for the NSDManager related implementation.
* Tested on a physical Android phone (screenshot attached below):
Screenshot_20220127-173149

### CODE REVIEW

File: examples/tv-casting-app/android/App/app/src/main/java/com/chip/casting/dnssd/DiscoveredNodeData.java [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+import java.util.List;
+import java.util.Map;
+
+public class DiscoveredNodeData {
+  private static final int MAX_IP_ADDRESSES = 5;
+  private static final int MAX_ROTATING_ID_LEN = 50;
+  private static final String KEY_DEVICE_NAME = "DN";
+  private static final String KEY_DEVICE_TYPE = "DT";
+
+
```
Comment (austinh0) [inline examples/tv-casting-app/android/App/app/src/main/java/com/chip/casting/dnssd/DiscoveredNodeData.java:17]:
```suggestion
```

File: examples/tv-casting-app/android/App/app/src/main/java/com/chip/casting/dnssd/DiscoveredNodeData.java [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+  private String deviceName;
+  private byte rotatingId[] = new byte[MAX_ROTATING_ID_LEN];
+  private int rotatingIdLen;
+  private short pairingHint;
+  private String pairingInstruction;
+  private short port;
+  private int numIPs;
+  private List<InetAddress> ipAddresses;
+
+  public DiscoveredNodeData() { }
```
Comment (austinh0) [inline examples/tv-casting-app/android/App/app/src/main/java/com/chip/casting/dnssd/DiscoveredNodeData.java:34]: Do we need this? When would we want to create an empty DiscoveredNodeData?

**Review by austinh0 — APPROVED** [review]

**Review by chrisdecenzo — APPROVED** [review]

File: examples/tv-casting-app/android/App/app/src/main/java/com/chip/casting/dnssd/CommissionerDiscoveryListener.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+import android.net.nsd.NsdManager;
+import android.net.nsd.NsdServiceInfo;
+import android.util.Log;
+import com.chip.casting.app.CastingContext;
+import com.chip.casting.util.GlobalCastingConstants;
+import java.util.ArrayList;
+import java.util.List;
+
+public class CommissionerDiscoveryListener implements NsdManager.DiscoveryListener {
```
[the remark this answers is not in the document]
Comment (sharadb-amazon) [reply examples/tv-casting-app/android/App/app/src/main/java/com/chip/casting/dnssd/CommissionerDiscoveryListener.java:11]: Thanks. Acquired the MulticastLock for a certain duration (eg: 10 sec) when sending the discovery request.

File: examples/tv-casting-app/android/App/app/src/main/java/com/chip/casting/dnssd/DiscoveredNodeData.java (hunk 2 above)
Comment (sharadb-amazon) [reply examples/tv-casting-app/android/App/app/src/main/java/com/chip/casting/dnssd/DiscoveredNodeData.java]: Removed it.

**Review by bzbarsky-apple — APPROVED** [review]
Rubberstamping.  Did not really read, fwiw, other than confirming only Android changes.

**Pull request merged**